### PR TITLE
filtering questions by tags securityConfig

### LINKED ISSUE #14: /getQuestionByTags needs authentication
{"timestamp":"2020-03-23T16:18:10.780+0000","status":401,"error":"Unauthorized","message":"Full authentication is required to access this resource","path":"/question/getQuestionsByTags"}

This service should have public access

### PR DESCRIPTION
**Pull request by Jakub-Mikolajczyk-pl**

### CODE REVIEW

File: src/main/java/pl/agasior/interviewprep/dto/GetQuestionsByTagsRequest.java [hunk 1]
```diff
@@ -11,6 +12,8 @@
 @Getter
 @Value
 public class GetQuestionsByTagsRequest {
+
+    @JsonFormat(with = JsonFormat.Feature.ACCEPT_SINGLE_VALUE_AS_ARRAY)
```
Comment (agasior093) [inline src/main/java/pl/agasior/interviewprep/dto/GetQuestionsByTagsRequest.java:18]: Could you exaplain why is this needed?

Comment (Jakub-Mikolajczyk-pl) [reply src/main/java/pl/agasior/interviewprep/dto/GetQuestionsByTagsRequest.java:18]: JSON parser in java, when got 1 object from front, didn't created it as array and got runtimeException when parsing

**Review by agasior093 — CHANGES_REQUESTED** [review]
1. I think that adding more tags to search criteria should reduce number of results, right now it just adds another subset of questions. It should be && instead of ||. 

2. That definitely needs integration test.

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @BeforeEach
+    @AfterAll
+    void clearDatabase() {
+        databasePreparer.clear();
+    }
+
+    @Nested
+    class ReadQuestion {
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: Not really needed if you got just one test

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+            final var command2 = CreateQuestionRequest.builder()
+                    .content("testContent")
+                    .answer("testAnswer")
+                    .tags(Set.of("testTag1", "testTag2", "testTag3")).build();
+
+            mockMvc.perform(requestFactory.createQuestion(command))
+                    .andDo(print())
+                    .andExpect(status().isOk())
+                    .andReturn();
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: You return resutl and you never use it.

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+            final var command = CreateQuestionRequest.builder()
+                    .content("testContent")
+                    .answer("testAnswer")
+                    .tags(Set.of("testTag1", "testTag2")).build();
+
+            final var command2 = CreateQuestionRequest.builder()
+                    .content("testContent")
+                    .answer("testAnswer")
+                    .tags(Set.of("testTag1", "testTag2", "testTag3")).build();
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: You repeat yourself while creating tag values, define constants instead.

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    .andExpect(status().isOk())
+                    .andReturn();
+
+            mockMvc.perform(requestFactory.createQuestion(command2))
+                    .andDo(print())
+                    .andExpect(status().isOk())
+                    .andReturn();
+
+            questionRepository.findByTags(new GetQuestionsByTagsRequest(Collections.singletonList("testTag1")))
+                            .forEach(quest -> Assertions.assertTrue(quest.getTags().contains("testTag1")));
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: I know its inside lambda and it can be short name, but quest is 'zadanie' :)

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+            Assertions.assertEquals(2, questionRepository.findByTags(
+                    new GetQuestionsByTagsRequest(Collections.singletonList("testTag1"))).size());
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: Didn't catch that before, repository is accepting request param? Its unnecessary bounding between persistance and API  layer, at this point repositroy can (and should) accept single Set<String> as parameter.

File: src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    void clearDatabase() {
+        databasePreparer.clear();
+    }
+
+    @Nested
+    class ReadQuestion {
+
+        @Test
+        void findQuestionsByTags() throws Exception {
+            final var user = UserDto.builder().username("admin").roles(Set.of(Role.Admin, Role.User)).build();
```
Comment (agasior093) [inline src/test/java/pl/agasior/interviewprep/services/question/QuestionReaderTest.java]: User creation isn't really part of this test and I bet its could be reused somewhere - maybe create factory class that will do that?

**Review by agasior093 — APPROVED** [review]
